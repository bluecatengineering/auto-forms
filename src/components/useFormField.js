import {useCallback, useContext} from 'react';

import FormContext from './FormContext';

export default (name) => {
	const {state, dispatch} = useContext(FormContext);
	const value = state.values[name];
	const rule = state.rules[name];
	const error = state.errors[name];
	const extra = state.extras[name];
	const setValue = useCallback(
		(value) => {
			dispatch({type: 'SET_VALUE', name, payload: value});
			if (rule) {
				dispatch({type: 'SET_ERROR', name, payload: rule(value)});
			}
		},
		[name, rule, dispatch]
	);

	const setError = useCallback((payload) => dispatch({type: 'SET_ERROR', name, payload}), [name, dispatch]);

	const setExtra = useCallback((payload) => dispatch({type: 'SET_EXTRA', name, payload}), [name, dispatch]);

	return {value, error, extra, setValue, setError, setExtra};
};
