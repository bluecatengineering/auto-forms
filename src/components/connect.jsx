import React, {useMemo} from 'react';
import PropTypes from 'prop-types';

import useFormField from './useFormField';

export default (Component, mapState) => {
	const Wrapper = ({name, ...props}) => {
		const state = useFormField(name);
		const newProps = {...props, ...mapState(state)};
		// eslint-disable-next-line react-hooks/exhaustive-deps
		return useMemo(() => <Component {...newProps} />, Object.values(newProps));
	};
	Wrapper.displayName = 'Form(' + (Component.displayName || Component.name) + ')';
	Wrapper.propTypes = {
		name: PropTypes.string.isRequired,
	};
	return Wrapper;
};
