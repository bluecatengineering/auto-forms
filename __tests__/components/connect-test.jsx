import React from 'react';
import {shallow} from 'enzyme';

import connect from '../../src/components/connect';
import useFormField from '../../src/components/useFormField';

jest.unmock('../../src/components/connect');

describe('connect', () => {
	it('returns a component which maps state', () => {
		const C = (...props) => <div {...props} />;
		const W = connect(
			C,
			({foo}) => ({bar: foo})
		);
		useFormField.mockReturnValue({foo: 'x'});
		const wrapper = shallow(<W id="test" name="bar" />);
		expect(wrapper.getElement()).toMatchSnapshot();
		expect(useFormField.mock.calls).toEqual([['bar']]);
	});
});
