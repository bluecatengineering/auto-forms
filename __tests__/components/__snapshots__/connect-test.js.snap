// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect returns a component which maps state 1`] = `
<C
  bar="x"
  id="test"
  name="bar"
/>
`;
